cmd/booster-http/piecehandler.go

```diff
@@ -0,0 +1,255 @@
+		return
+	}
+
+	setHeaders(w, pieceCid)
+	serveContent(w, r, content)
+
+	stats.Record(ctx, metrics.HttpPieceByCid200ResponseCount.M(1))
+	stats.Record(ctx, metrics.HttpPieceByCidRequestDuration.M(float64(time.Since(startTime).Milliseconds())))
+}
+
+func (s *HttpServer) getPieceContent(ctx context.Context, pieceCid cid.Cid) (io.ReadSeeker, error) {
+	// Get the deals for the piece
+	pieceDeals, err := s.api.GetPieceDeals(ctx, pieceCid)
+	if err != nil {
+		return nil, fmt.Errorf("getting sector info for piece %s: %w", pieceCid, err)
+	}
+
+	// Get the first unsealed deal
+	di, err := s.unsealedDeal(ctx, pieceCid, pieceDeals)
+	if err != nil {
+		return nil, fmt.Errorf("getting unsealed CAR file: %w", err)
+	}
+
+	// Get the raw piece data from the sector
+	pieceReader, err := s.api.UnsealSectorAt(ctx, di.MinerAddr, di.SectorID, di.PieceOffset.Unpadded(), di.PieceLength.Unpadded())
+	if err != nil {
+		return nil, fmt.Errorf("getting raw data from sector %d: %w", di.SectorID, err)
+	}
+
+	return pieceReader, nil
+}
+
+func isGzipped(res http.ResponseWriter) bool {
+	switch res.(type) {
+	case *gziphandler.GzipResponseWriter, gziphandler.GzipResponseWriterWithCloseNotify:
+		// there are conditions where we may have a GzipResponseWriter but the
+		// response will not be compressed, but they are related to very small
+		// response sizes so this shouldn't matter (much)
+		return true
+	}
+	return false
+}
+
+func setHeaders(w http.ResponseWriter, pieceCid cid.Cid) {
+	w.Header().Set("Vary", "Accept-Encoding")
+	etag := `"` + pieceCid.String() + `"` // must be quoted
+	if isGzipped(w) {
+		etag = etag[:len(etag)-1] + ".gz\""
+	}
+	w.Header().Set("Etag", etag)
+	w.Header().Set("Content-Type", "application/piece")
+	w.Header().Set("Cache-Control", "public, max-age=29030400, immutable")
+}
+
+func serveContent(res http.ResponseWriter, req *http.Request, content io.ReadSeeker) {
+	// http.ServeContent ignores errors when writing to the stream, so we
+	// replace the writer with a class that watches for errors
+	res = newPieceAccountingWriter(res, toLoggingResponseWriter(res))
+
+	// Note that the last modified time is a constant value because the data
+	// in a piece identified by a cid will never change.
+
+	if req.Method == "HEAD" {
+		// For an HTTP HEAD request ServeContent doesn't send any data (just headers)
+		http.ServeContent(res, req, "", time.Time{}, content)
+		return
+	}
+
+	// Send the content
+	http.ServeContent(res, req, "", lastModified, content)
+}
+
+// isNotFoundError falls back to checking the error string for "not found".
+// Unfortunately we can't always use errors.Is() because the error might
+// have crossed an RPC boundary.
+func isNotFoundError(err error) bool {
+	if errors.Is(err, ErrNotFound) {
+		return true
+	}
+	if errors.Is(err, datastore.ErrNotFound) {
+		return true
+	}
+	if errors.Is(err, retrievalmarket.ErrNotFound) {
+		return true
+	}
+	return strings.Contains(strings.ToLower(err.Error()), "not found")
```

Comment:
Subjective: assuming we want to avoid a sting of `||`s, probably more readable with a `switch` statement?
```suggestion
func isNotFoundError(err error) bool {
	switch {
	case errors.Is(err, ErrNotFound),
		errors.Is(err, datastore.ErrNotFound),
		errors.Is(err, retrievalmarket.ErrNotFound),
		strings.Contains(strings.ToLower(err.Error()), "not found"):
		return true
	default:
		return false
	}
	}
```